geps/gep-1713/index.md

```diff
@@ -0,0 +1,433 @@
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=64
+	//
+	// Note: this is the same ListenerStatus type in the GatewayStatus struct
+	Listeners []ListenerStatus `json:"listeners,omitempty"`
+
+	// Conditions describe the current conditions of the ListenerSet.
+	//
+	// Implementations should prefer to express ListenerSet conditions
+	// using the `GatewayConditionType` and `GatewayConditionReason`
+	// constants so that operators and tools can converge on a common
+	// vocabulary to describe Gateway state.
+	//
+	// Known condition types are:
+	//
+	// * "Accepted"
+	// * "Programmed"
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MaxItems=8
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// ParentGatewayReference identifies an API object including its namespace,
+// defaulting to Gateway.
+type ParentGatewayReference struct {
+	// Group is the group of the referent.
+	//
+	// +optional
+	// +kubebuilder:default="gateway.networking.k8s.io"
+	Group *Group `json:"group"`
+
+	// Kind is kind of the referent. For example "Gateway".
+	//
+	// +optional
+	// +kubebuilder:default=Gateway
+	Kind *Kind `json:"kind"`
+
+	// Name is the name of the referent.
+	Name ObjectName `json:"name"`
+}
+```
+
+### YAML
+
+The following example shows a `Gateway` with an HTTP listener and two child HTTPS `ListenerSets` with unique hostnames and certificates.
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1
+kind: Gateway
+metadata:
+  name: parent-gateway
+spec:
+  gatewayClassName: example
+  listeners:
+  - name: foo
+    hostname: foo.com
+    protocol: HTTP
+    port: 80
+---
+apiVersion: gateway.networking.k8s.io/v1alpha1
+kind: ListenerSet
+metadata:
+  name: first-workload-listeners
+spec:
+  parentRefs:
+  - name: parent-gateway
+    kind: Gateway
+    group: gateway.networking.k8s.io
+  listeners:
+  - name: first
+    hostname: first.foo.com
+    protocol: HTTPS
+    port: 443
+    tls:
+      mode: Terminate
+      certificateRefs:
+      - kind: Secret
+        group: ""
+        name: first-workload-cert # Provisioned via HTTP01 challenge
+---
+apiVersion: gateway.networking.k8s.io/v1alpha1
+kind: ListenerSet
+metadata:
+  name: second-workload-listeners
+spec:
+  parentRefs:
+  - name: parent-gateway
+    kind: Gateway
+    group: gateway.networking.k8s.io
+  listeners:
+  - name: second
+    hostname: second.foo.com
+    protocol: HTTPS
+    port: 443
+    tls:
+      mode: Terminate
+      certificateRefs:
+      - kind: Secret
+        group: ""
+        name: second-workload-cert # Provisioned via HTTP01 challenge
+```
+## Semantics
+
+### Gateway Changes
+
+When an implementation supports `ListenerSets` `Gateways` MUST allow the list of listeners to be empty. Thus the present `minItems=1` constraint on the listener list will be removed. This allows implementations to avoid security, cost etc. concerns with having dummy listeners.
```

Comment:
Other folks feel pretty strongly about this

see: https://github.com/kubernetes-sigs/gateway-api/pull/3213#discussion_r1697321868

Reply:
I agree that dummy listeners are not ergonomic, but I think they're a necessity until we're more sure that this approach is going to move ahead.

Changing Gateway to allow empty Listeners is not a breaking _API_ change, since you can take a Gateway that existed before the change, change it to the new schema, then change it back with no loss of information.

It's a big behavior change though - especially since it complicates acceptance quite a bit.

Should a Gateway with no Listeners be accepted? Currently the answer is _no_, because that means that there _are_ Listeners present, but that none of those Listeners are valid. If we make Listeners optional and able to be empty, presumably we want a Gateway with _empty_ Listeners to be treated differently than a Gateway with _zero invalid_ Listeners and be accepted? If so, we'll have to _carefully_ write down the rules there.

My pretty strong preference here is for ListenerSets to actually be an experiment, and require _zero_ changes to the GA objects until we are confident that we will proceed. If we change the GA objects to allow empty Listeners, that's it, we can _never_ walk that back, as walking it back _is_ a breaking API change.

Reply:
That does mean that, for the initial release of ListenerSets, we'll need to document some form of explicit dummy Listener workaround, with clear explanation that this is for _initial experimentation only_, and if decide to move forward after (probably) one release, we'll make changes to make Listeners optional.

Reply:
added a note about this in the GEP